self-core-impl/src/main/java/com/selfxdsd/core/ConditionalJsonResources.java

```diff
@@ -142,48 +143,51 @@ private Resource conditionalGet(
         final Supplier<Map<String, List<String>>> headers
     ) {
         final Resource resource;
-        final String etag = this.jsonStorage.getEtag(uri);
-        if (etag != null) {
+        final ConditionalResource stored = this.jsonStorage.getResource(uri);
+        if (stored != null) {
             final Resource remoteResource = this.delegate
-                .get(uri, this.ifNoneMatch(headers, etag));
+                .get(uri, this.ifNoneMatch(headers, stored.etag()));
             final int status = remoteResource.statusCode();
 
             if (status == HttpURLConnection.HTTP_NOT_MODIFIED) {
-                final String cachedBody = this.jsonStorage.getResourceBody(uri);
-                if (cachedBody != null) {
-                    LOG.debug(
-                        "Remote resource body for {} was not modified."
-                            + " Getting the resource body from json storage.",
-                        uri
-                    );
-                    resource = remoteResource
-                        .newBuilder()
-                        .status(HttpURLConnection.HTTP_OK)
-                        .body(cachedBody)
-                        .build();
-                } else {
-                    LOG.debug(
-                        "Remote resource for {} was not modified"
-                            + " but resource body is missing from json storage."
-                            + " Getting the resource from remote.",
-                        uri
-                    );
-                    resource = this.delegate.get(uri, headers);
-                    this.storeInCache(uri, resource);
-                }
+                LOG.debug(
+                    "Remote resource body for {} was not modified."
+                        + " Getting the resource body from json storage.",
+                    uri
+                );
+                resource = stored;
             } else {
                 LOG.debug(
                     "Remote resource body for {} was modified or "
                         + " has an unexpected status code.",
                     uri
                 );
-                resource = remoteResource;
-                this.storeInCache(uri, resource);
+                final ConditionalResource newConditional = ConditionalResource
+                    .fromResource(uri, remoteResource);
+                if (newConditional != null) {
```

Comment:
@criske here we will just have ``if(cached.eTag() != null)``

Reply:
@amihaiemil Doesn't make sense to have CachedResource `eTag` null, that's why is cached in the first place :D